### PR TITLE
Implement the `enum` function

### LINKED ISSUE #3681: Support reading enums from PHP 8.1
Hi,

Since PHP 8.1 we can use Enums. Which in my opinion are a great asset to PHP.
But unfortunate there is not an elegant way of retrieving values inside Twig templates.

I currently use this in my Symfony project;

```twig
{{ constant('App\\...\\Enum::Key').value }}
```

Which will return the value of the key in de the Enum.

Maybe it's good to add some specific functions for this in the Twig core?

Regards, Bert

### PR DESCRIPTION
**Pull request by nicolas-grekas**
Fixes #3681
Requires #4353

### CODE REVIEW

**Comment by fabpot** [comment]
@nicolas-grekas Can you rebase?

File: doc/functions/enum.rst [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+
+``enum`` gives access to enums:
+
+.. code-block:: twig
+
+    {# display one specific case of a backed enum #}
+    {{ enum('App\\MyEnum').SomeCase.value }}
+
+    {# get all cases of an enum #}
+    {% enum('App\\MyEnum').cases() %}
```
Comment (fabpot) [inline doc/functions/enum.rst:16]: Should we deprecate `enum_cases()`?

Comment (stof) [reply doc/functions/enum.rst:16]: I would not deprecated it. `enum_cases` compiles to optimized code when the enum class name is a string literal, avoiding any dot operator resolution at runtime.

Comment (fabpot) [reply doc/functions/enum.rst:16]: 👍

**Comment by fabpot** [comment]
Thank you @nicolas-grekas.

**Pull request merged**